File: frontend/src/lib/trip.ts
import { decodePolyline } from "./routing";

export interface TripData {
  id: string;
  creatorID: string;
  creatorName: string;
  originLat: number;
  originLng: number;
  originName: string;
  destLat: number;
  destLng: number;
  destName: string;
  routeGeometry: string;
  distanceMeters: number;
  durationSeconds: number;
  status: string;
  participants: string[];
  startedAt: number | null;
  endedAt: number | null;
  createdAt: number;
}

export function parseRouteCoordinates(routeGeometry: string): [number, number][] {
  if (!routeGeometry) return [];
  try {
    const parsed = JSON.parse(routeGeometry);
    if (Array.isArray(parsed) && parsed.length > 0 && Array.isArray(parsed[0])) {
      return parsed as [number, number][];
    }
  } catch {}
  return decodePolyline(routeGeometry);
}

export async function fetchActiveTrip(groupID: string): Promise<TripData | null> {
  const res = await fetch(`/api/trip/${encodeURIComponent(groupID)}`);
  if (!res.ok) return null;
  const text = await res.text();
  if (text === "null" || !text) return null;
  return JSON.parse(text);
}

File: backend/internal/model/trip.go
package model

// Trip represents a shared trip between group members.
type Trip struct {
	ID              string   `json:"id"`
	GroupID         string   `json:"groupID"`
	CreatorID       string   `json:"creatorID"`
	CreatorName     string   `json:"creatorName"`
	OriginLat       float64  `json:"originLat"`
	OriginLng       float64  `json:"originLng"`
	OriginName      string   `json:"originName"`
	DestLat         float64  `json:"destLat"`
	DestLng         float64  `json:"destLng"`
	DestName        string   `json:"destName"`
	RouteGeometry   string   `json:"routeGeometry"`
	DistanceMeters  float64  `json:"distanceMeters"`
	DurationSeconds float64  `json:"durationSeconds"`
	Status          string   `json:"status"`
	Participants    []string `json:"participants"`
	StartedAt       *int64   `json:"startedAt,omitempty"`
	EndedAt         *int64   `json:"endedAt,omitempty"`
	CreatedAt       int64    `json:"createdAt"`
}

const (
	TripStatusPlanning  = "planning"
	TripStatusActive    = "active"
	TripStatusCompleted = "completed"
)
